packages/types/types/src/record.ts

```diff
@@ -3,45 +3,102 @@
  * Licensed under the MIT License.
  */
 
+/** An observable key/value collection, such as a Record or Map. */
+export interface IShapeProducer<TKey> {
+    /**
+     * Acquire a reader for this producer's values and implicitly subscribe the consumer
+     * to value change notifications.
+     *
+     * @param consumer - The consumer to be notified of value changes.
+     */
+    open(consumer: IShapeConsumer<TKey>): IShapeReader<TKey>;
+
+    /**
+     * Unsubscribe the given 'consumer' from this producer's change notifications.
+     *
+     * @param consumer - The consumer to unregister from the producer.
+     */
+    close(consumer: IShapeConsumer<TKey>): void;
+}
+
+/** A consumer of change notifications for a key/value collection, such as a Record or Map. */
+export interface IShapeConsumer<TKey> {
+    keyChanged(key: TKey, producer: IShapeProducer<TKey>): void;
+}
+
+/** Capability to read values of a key/value collection. */
+export interface IShapeReader<TKey> {
+    keys(): IterableIterator<TKey>;
+
+    has(key: TKey): boolean;
+
+    readonly size: number;
+
+    /**
+     * A reference to the underlying producer that provides the shape for this reader,
+     * or undefined if the producer is immutable.
+     */
+    readonly producer?: IShapeProducer<TKey>;
+}
+
+export interface IShapeWriter<TKey> {
+    delete(key: TKey): void;
+    clear(): void;
+}
+
 /** An observable key/value collection, such as a Record or Map. */
 export interface IProducer<T> {
     /**
      * Acquire a reader for this producer's values and implicitly subscribe the consumer
      * to value change notifications.
-     * 
+     *
      * @param consumer - The consumer to be notified of value changes.
      */
     open(consumer: IConsumer<T>): IReader<T>;
 
     /**
      * Unsubscribe the given 'consumer' from this producer's change notifications.
-     * 
+     *
      * @param consumer - The consumer to unregister from the producer.
      */
     close(consumer: IConsumer<T>): void;
 }
 
+/** A consumer of change notifications for a key/value collection, such as a Record or Map. */
+export interface IConsumer<T> {
+    keyChanged<K extends keyof T>(key: K, producer: IProducer<T>): void;
+}
+
 /** Capability to read values of a key/value collection. */
 export interface IReader<T> {
```

Comment:
I think we've been implicitly assuming that `T` is object-like, rather than a primitive. I'm not sure it is worth putting this in the constraint, it depends on the dev experience. I think it should be left as-is, but I'd be interested to see if we find any practical reasons for/against.

Reply:
Oh yeah...  you're right.

One consideration: right now I'm assuming 'Object.keys(instanceOfT)' and 'keyof T' contain the same  set of keys, but that's only true if 'T' is a plain object.  For something like a string, 'keyof T' will contain a lot of extra stuff like 'toString()', 'indexOf()', 'toLowerCase()', etc.